### PR TITLE
Video uploading,thumbnail & Refactoring

### PR DESCRIPTION
**Pull request by moizansari22**
Test assignment working fine 

1. seed functionality implemented for categories
2. index page consist of list includes video title, video thumbnail (256x256), video category 
3. mouseover functionality implemented using JQuery but it is in view section.
4. At backend upload button working fine 
5. validation for supported file type and max size implemented 
6. devise is implemented at initial run of application 
7. title , introduction , video & thumbnail added to database : video is stored using activestorage
8. three thumbnails doesn't works will check it tomorrow.

Screenshot 2022-10-11 at 6 50 12 PM

### CODE REVIEW

File: app/views/videos/_video.html.erb [hunk 1]
```diff
@@ -1,14 +1,18 @@
-<div id="<%= dom_id video %>">
+<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.6.0/jquery.min.js"></script>
```
Comment (MuhammadRaza29) [inline app/views/videos/_video.html.erb]: You have already added` jquery-rails` gem then you are manually adding query here?

File: app/views/videos/_video.html.erb [hunk 2]
```diff
@@ -1,14 +1,18 @@
...
+<div align="center" id="<%= dom_id video %>">
```
Comment (MuhammadRaza29) [inline app/views/videos/_video.html.erb:2]: Why you are assigning video object as id?

File: db/migrate/20221011111744_drop_categories_table.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class DropCategoriesTable < ActiveRecord::Migration[7.0]
```
Comment (MuhammadRaza29) [inline db/migrate/20221011111744_drop_categories_table.rb]: Why you are dropping categories table?

File: db/seeds.rb [hunk 4]
```diff
@@ -5,3 +5,8 @@
 #
 #   movies = Movie.create([{ name: "Star Wars" }, { name: "Lord of the Rings" }])
 #   Character.create(name: "Luke", movie: movies.first)
+puts "Seeding..."
+Category.create(name: 'Excersize')
+Category.create(name: 'Education')
+Category.create(name: 'Recipe')
```
Comment (MuhammadRaza29) [inline db/seeds.rb]: Kindly use for loop to create categories.

File: app/models/category.rb [hunk 5]
```diff
@@ -0,0 +1,3 @@
+class Category < ApplicationRecord
+    has_one :video
```
Comment (MuhammadRaza29) [inline app/models/category.rb]: How category can have one video?

File: app/views/videos/_video.html.erb [hunk 6]
```diff
@@ -1,14 +1,18 @@
...
     <strong>Title:</strong>
     <%= video.title %>
   </p>
-
-  <p>
-    <strong>Introduction:</strong>
-    <%= video.introduction %>
-  </p>
-
+  <%= video_tag(url_for(video.video) , size: "650x400",:controls => true) %>
```
Comment (MuhammadRaza29) [inline app/views/videos/_video.html.erb]: What's the reason of specifying `size: "650x400"`?

File: app/views/videos/_video.html.erb (hunk 1 above)
Comment (moizansari22) [reply app/views/videos/_video.html.erb]: resolved this issue .

File: db/migrate/20221011111744_drop_categories_table.rb (hunk 3 above)
Comment (moizansari22) [reply db/migrate/20221011111744_drop_categories_table.rb]: deleted the migration and added new migration to drop 'introduction' from videos table

File: app/models/category.rb (hunk 5 above)
Comment (moizansari22) [reply app/models/category.rb]: category have many videos & videos belongs to category. issue resolved

File: app/views/videos/_video.html.erb (hunk 6 above)
Comment (moizansari22) [reply app/views/videos/_video.html.erb]: the reason to add the size was when user will upload a new video . the video may be of any size so, to see all the videos in equal size at index page we use the method size .

File: db/seeds.rb (hunk 4 above)
Comment (moizansari22) [reply db/seeds.rb]: used loop and followed dry principle of rails .

File: app/controllers/videos_controller.rb [hunk 7]
```diff
@@ -63,7 +60,10 @@ def destroy
   private
     # Use callbacks to share common setup or constraints between actions.
     def set_video
-      @video = Video.find(params[:id])
+      @video = Video.find_by(:id)
```
Comment (MuhammadRaza29) [inline app/controllers/videos_controller.rb]: @moizansari22 Did you test it?

Comment (moizansari22) [reply app/controllers/videos_controller.rb]: again implemented and tested and pushed the code to a new commit named as "Latest incorporated feedback"

File: app/controllers/videos_controller.rb [hunk 8]
```diff
@@ -63,11 +60,14 @@ def destroy
   private
     # Use callbacks to share common setup or constraints between actions.
     def set_video
-      @video = Video.find(params[:id])
+      @video = Video.find_by(id: params[:id])
+      if @video.nil?
+        flash[:alert] = "video not avalible"
+        redirect_to "/videos#index"
```
Comment (MuhammadRaza29) [inline app/controllers/videos_controller.rb]: @moizansari22 Kindly use the named route.
https://guides.rubyonrails.org/routing.html#controller-namespaces-and-routing

File: app/controllers/videos_controller.rb [hunk 9]
```diff
@@ -63,11 +60,14 @@ def destroy
...
+      end
     end
-
     # Only allow a list of trusted parameters through.
     def video_params
-      params.require(:video).permit(:title, :introduction , :video , :category)
+      params.require(:video).permit(:title , :video , :category_id)
```
Comment (MuhammadRaza29) [inline app/controllers/videos_controller.rb]: @moizansari22 
```suggestion
      params.require(:video).permit(:title, :video , :category_id)
```

File: app/controllers/videos_controller.rb (hunk 8 above)
Comment (MuhammadRaza29) [inline app/controllers/videos_controller.rb]: Kindly use the named route.
https://guides.rubyonrails.org/routing.html#controller-namespaces-and-routing

Comment (moizansari22) [reply app/controllers/videos_controller.rb]: done redirected to root path

File: app/controllers/videos_controller.rb [hunk 10]
```diff
@@ -1,12 +1,8 @@
 class VideosController < ApplicationController
   before_action :set_video, only: %i[ show edit update destroy ]
-  before_action :authenticate_user! 
-  skip_before_action :verify_authenticity_token
-  def home
-  end
   # GET /videos or /videos.json
   def index
-    @videos = Video.all
+    @videos = Video.all.paginate(:page => params[:page], :per_page => 10)
```
Comment (MuhammadRaza29) [inline app/controllers/videos_controller.rb]: Kindly use new hash syntax.

File: app/controllers/videos_controller.rb [hunk 11]
```diff
@@ -63,11 +60,14 @@ def destroy
   private
     # Use callbacks to share common setup or constraints between actions.
     def set_video
-      @video = Video.find(params[:id])
+      @video = Video.find_by(id: params[:id])
+      if @video.nil?
+        flash[:alert] = "video not avalible"
+        redirect_to :root
```
Comment (MuhammadRaza29) [inline app/controllers/videos_controller.rb]:
```suggestion
        redirect_to root_path
```

File: app/models/category.rb [hunk 12]
```diff
@@ -0,0 +1,3 @@
+class Category < ApplicationRecord
+  has_many :video
```
Comment (MuhammadRaza29) [inline app/models/category.rb]:
```suggestion
  has_many :videos
```

**Review by MuhammadRaza29 — APPROVED** [review]
